x-pack/legacy/plugins/lens/public/editor_frame_service/editor_frame/config_panel_wrapper.tsx

```diff
@@ -174,85 +190,329 @@ function LayerPanels(
+  const layerDatasource = props.datasourceMap[datasourceId];
 
-  return (
-    <EuiPanel className="lnsConfigPanel__panel" paddingSize="s">
-      <EuiFlexGroup gutterSize="s" alignItems="flexStart" responsive={false}>
-        <EuiFlexItem grow={false}>
-          <LayerSettings
-            layerId={layerId}
-            layerConfigProps={layerConfigProps}
-            activeVisualization={activeVisualization}
-          />
-        </EuiFlexItem>
+  const layerDatasourceDropProps = {
+    layerId,
+    dragDropContext,
+    state: layerDatasourceState,
+    setState: (newState: unknown) => {
+      props.updateDatasource(datasourceId, newState);
+    },
+  };
+
+  const layerDatasourceConfigProps = {
+    ...layerDatasourceDropProps,
+    frame: props.framePublicAPI,
+    dateRange: props.framePublicAPI.dateRange,
+  };
 
-        {datasourcePublicAPI && (
-          <EuiFlexItem className="eui-textTruncate">
-            <NativeRenderer
-              render={datasourcePublicAPI.renderLayerPanel}
-              nativeProps={{ layerId }}
+  const [popoverState, setPopoverState] = useState<{
+    isOpen: boolean;
+    openId: string | null;
+  }>({
+    isOpen: false,
+    openId: null,
+  });
+
+  const { dimensions } = activeVisualization.getLayerOptions(layerVisualizationConfigProps);
+  const isEmptyLayer = !dimensions.some(d => d.accessors.length > 0);
+
+  function wrapInPopover(id: string, trigger: React.ReactElement, panel: React.ReactElement) {
+    return (
+      <EuiPopover
+        isOpen={popoverState.isOpen && popoverState.openId === id}
+        closePopover={() => {
+          setPopoverState({ isOpen: false, openId: null });
+        }}
+        button={trigger}
+        display="block"
+        anchorPosition="leftUp"
+        withTitle
+        panelPaddingSize="s"
+      >
+        {panel}
+      </EuiPopover>
+    );
+  }
+
+  return (
+    <ChildDragDropProvider {...dragDropContext}>
+      <EuiPanel className="lnsConfigPanel__panel" paddingSize="s">
+        <EuiFlexGroup gutterSize="s" alignItems="flexStart" responsive={false}>
+          <EuiFlexItem grow={false}>
+            <LayerSettings
+              layerId={layerId}
+              layerConfigProps={layerVisualizationConfigProps}
+              activeVisualization={activeVisualization}
             />
           </EuiFlexItem>
-        )}
-      </EuiFlexGroup>
 
-      <EuiSpacer size="s" />
+          {datasourcePublicAPI && (
+            <EuiFlexItem className="eui-textTruncate">
+              <NativeRenderer
+                render={datasourcePublicAPI.renderLayerPanel}
+                nativeProps={{ layerId }}
+              />
+            </EuiFlexItem>
+          )}
+        </EuiFlexGroup>
 
-      <NativeRenderer
-        render={activeVisualization.renderLayerConfigPanel}
-        nativeProps={layerConfigProps}
-      />
+        <EuiSpacer size="s" />
+
+        {dimensions.map((dimension, index) => {
+          const newId = generateId();
```

Comment:
For reviewers: This call to `generateId()` is important. I will write some other comments here explaining why.